### PR TITLE
[Dependency Update] Dependency update doc

### PR DESCRIPTION
**Pull request by stu1130**
## Description ##
document dependency update

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by karan6181** [comment]
@mxnet-label-bot add [Doc, pr-awaiting-review]

**Comment by vandanavk** [comment]
@marcoabreu @perdasilva @aaronmarkham for review

**Comment by karan6181** [comment]
@marcoabreu @perdasilva @aaronmarkham Gentle Ping!

File: tools/dependencies/README.md [hunk 1]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
 ```
 --without-libidn2
 ``` 
+
+***
+
+# Dependency Update Runbook
+
+MXNet is built on top of many dependencies. Managing those dependencies could be a big headache. This goal of this document is to give a overview of those dependencies and how to upgrade when new version of those are rolled out.
```
Comment (aaronmarkham) [inline tools/dependencies/README.md]:
```suggestion
MXNet is built on top of many dependencies. Managing these dependencies can be a big headache. This goal of this document is to give a overview of MXNet's dependencies and how to upgrade them when new versions are rolled out.
```

File: tools/dependencies/README.md [hunk 2]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+    python3-pip \
+    automake \
+    pkg-config \
+    openjdk-8-jdk
+    
+# CUDA installation 
+# Take CUDA 10 for example
+wget https://developer.nvidia.com/compute/cuda/10.0/Prod/local_installers/cuda_10.0.130_410.48_linux
+chmod +x cuda_10.0.130_410.48_linux && sudo ./cuda_10.0.130_410.48_linux
+# Installation except:
```
Comment (aaronmarkham) [inline tools/dependencies/README.md]: What does this mean?

File: tools/dependencies/README.md [hunk 3]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+## How to update them?
+
+### MKL, MKLDNN
+
+@pengzhao-intel (https://github.com/apache/incubator-mxnet/commits?author=pengzhao-intel) and his team are tracking and updating these versions. Kudos to them!
+
+### CUDA, cuDNN, NCCL
+#### 1. Environment Setup
+```
+# Take Ubuntu 16.04 for example
```
Comment (aaronmarkham) [inline tools/dependencies/README.md]: You could move this out of the code block and introduce what is going on here.

File: tools/dependencies/README.md [hunk 4]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+# https://developer.nvidia.com/nccl/nccl2-download-survey
+# Register with NVIDIA and download nccl-repo-ubuntu1604-2.4.2-ga-cuda10.0_1-1_amd64.deb
+sudo dpkg -i nccl-repo-ubuntu1604-2.4.2-ga-cuda10.0_1-1_amd64.deb
+sudo apt-key add /var/nccl-repo-2.4.2-ga-cuda10.0/7fa2af80.pub
+sudo apt update
+sudo apt install libnccl2 libnccl-dev
+# we will check the NCCL version later
+```
+#### 2. Build 
+```
```
Comment (aaronmarkham) [inline tools/dependencies/README.md]: Introduce this section...

File: tools/dependencies/README.md [hunk 5]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+nosetests --verbose tests/python/gpu/test_nccl.py
+# test_nccl.test_nccl_pushpull ... NCCL version 2.4.2+cuda10.0
+# ok
+# ----------------------------------------------------------------------
+# Ran 1 test in 67.666s
+
+OK
+```
+#### 3. Performance Sanity Check
+We will test against 3 basic models
```
Comment (aaronmarkham) [inline tools/dependencies/README.md]: Use periods after each of the statements in this section.

File: tools/dependencies/README.md [hunk 6]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+###### ResNet50 with ImageNet
+```
+# Download the ImageNet on http://image-net.org/download and make record file
+# Install prerequisite package
+pip2 install psutil --user
```
Comment (apeforest) [inline tools/dependencies/README.md]: Why pip2?

File: tools/dependencies/README.md [hunk 7]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+2. Build PyPi package
+```
+# Update the dependency under tools/dependencies, then
+tools/staticbuild/build.sh mkl pip
+
+# Wait for 10 - 30 mins, you will find libmxnet.so under the incubator-mxnet/lib
+
+# Install python frontend
+cd python
+pip3 install -e . --pre
```
Comment (apeforest) [inline tools/dependencies/README.md]: maybe just pip? Let users to pick the right version (maybe thorugh viurtualenv)

File: tools/dependencies/README.md [hunk 8]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+    automake \
+    pkg-config \
+    openjdk-8-jdk
+    
+# CUDA installation 
+# Take CUDA 10 for example
+wget https://developer.nvidia.com/compute/cuda/10.0/Prod/local_installers/cuda_10.0.130_410.48_linux
+chmod +x cuda_10.0.130_410.48_linux && sudo ./cuda_10.0.130_410.48_linux
+# Installation except:
+# Install NVIDIA Accelerated Graphics Driver for Linux-x86_64 410.48?
```
Comment (apeforest) [inline tools/dependencies/README.md]: I guess you can just point to the NVIDIA CUDA driver installation page directly. I also found installing from deb (network) is very simple. Just 4 lines according to: https://developer.nvidia.com/cuda-downloads?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1604&target_type=debnetwork

File: tools/dependencies/README.md (hunk 2 above)
Comment (stu1130) [reply tools/dependencies/README.md]: Type fixed

File: tools/dependencies/README.md [hunk 9]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+# Make sure you add new makefile and right debs CUDA uses on the website
+# http://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/
+
+# Build PyPi package
+tools/staticbuild/build.sh cu100mkl pip
+
+# Wait for 10 - 30 mins, you will find libmxnet.so under the incubator-mxnet/lib
+
+# Install python frontend
+cd python
```
Comment (apeforest) [inline tools/dependencies/README.md]: This can be simplified by pip install -e python .

File: tools/dependencies/README.md (hunk 6 above)
Comment (stu1130) [reply tools/dependencies/README.md]: our benchmark script is written in python2

File: tools/dependencies/README.md [hunk 10]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+pip install -e . --pre
+# Test MXNet
+>>> import mxnet as mx
+>>> mx.nd.ones((2, 5) ctx=mx.gpu(0))
```
Comment (apeforest) [inline tools/dependencies/README.md]: Maybe you can replace this test using the mxnet.runtime.feature_list()

File: tools/dependencies/README.md [hunk 11]
```diff
@@ -40,3 +40,349 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+```
+# Download the ImageNet on http://image-net.org/download and make record file
+# Install prerequisite package
+pip2 install psutil --user
+pip2 install pandas --upgrade --user
+pip install gluoncv==0.2.0b20180625
+# Clone the testing script
+git clone https://github.com/rahul003/deep-learning-benchmark-mirror.git
+# command 
+python2 benchmark_runner.py --framework mxnet --metrics-policy metrics_parameters_images_top_1 --task-name metrics_parameters_images_top_1 --metrics-suffix test --num-gpus 8 --command-to-execute 'python mxnet_benchmark/train_imagenet.py --use-rec --batch-size 128 --dtype float32 --num-data-workers 40 --num-epochs 3 --gpus 0,1,2,3,4,5,6,7 --lr 0.4 --warmup-epochs 5 --last-gamma --mode symbolic --model resnet50_v1b --rec-train /home/ubuntu/data/train-passthrough.rec --rec-train-idx /home/ubuntu/data/train-passthrough.idx --rec-val /home/ubuntu/data/val-passthrough.rec --rec-val-idx /home/ubuntu/data/val-passthrough.idx' --data-set data
```
Comment (apeforest) [inline tools/dependencies/README.md]: This benchmark AI flow is outdated and I think it's also not public. Why not just run training script from example directly?

Comment (apeforest) [reply tools/dependencies/README.md]: Also maybe use mnist as an example here. ImageNet is too big.

File: tools/dependencies/README.md [hunk 12]
```diff
@@ -40,3 +40,301 @@ This issue appeared in the OSX build with XCode version 8.0 above (reproduced on
...
+
+# Dependency Update Runbook
+
+MXNet is built on top of many dependencies. Managing these dependencies could be a big headache. This goal of this document is to give a overview of those dependencies and how to upgrade when new version of those are rolled out.
+
+## Overview
+
+The dependencies could be categorized by several groups: BLAS libraries, CPU-based performance boost library, i.e. MKLDNN and GPU-based performance boosting library including CUDA, cuDNN, NCCL. and others including OpenCV, Numpy, S3-related, PS-lite dependencies. The list below shows all the dependencies and their version. Except for CUDA, cuDNN, NCCL which the user is required to install on their environments, we statically link those dependencies into libmxnet.so when we build PyPi package. By doing this, the user can take advantage of these dependencies without being worry about it.
+
+| Dependencies  | MXNet Version |
```
Comment (apeforest) [inline tools/dependencies/README.md]: Is this MXNet version or just version of each dependency?

**Comment by aaronmarkham** [comment]
@stu1130 Did you want to make any further updates to this?

**Comment by stu1130** [comment]
@aaronmarkham I think it's good to go

**Pull request merged**